### PR TITLE
use try/finally to rollback or commit

### LINKED ISSUE #8997: Transactions not rolled back when fatal errors are thrown in PHP7
This is a (multiple allowed):
* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.2.10
* Platform and Target: MySQL, PHP7, Apache 2.4

### What you did

Used transactional so that errors would trigger a rollback.

### Expected Behavior

Callback function can throw an error that triggers the rollback.

### Actual Behavior

Callback throws an error, but transaction does not rollback. Execution block is escaped, and the transaction is not closed.

The following is a catchable error.

```
    try {
        $foo = array();
        $this->$foo(); // $foo is not a string.
   } catch(Error $er) {
         // catches error "Function name must be a string"
         Log::error($er->getMessage());
   }
```

What happen to me today.

```
      try {
           // bla bla bla...
           $this->connection()->transactional(function() use ($foo) {
                     $this->$foo(); // throws error
           });
     } finally {
           $this->LogTable->save($record);
           // the above is still inside a transaction. Transactional never rollback or commit.
     }
```

The issue exists here in the code for `Connection.php`

```
    public function transactional(callable $callback)
    {
        $this->begin();

        try {
            $result = $callback($this);
        } catch (Exception $e) {
            $this->rollback();
            throw $e;
        }

        if ($result === false) {
            $this->rollback();
            return false;
        }

        $this->commit();
        return $result;
    }
```

The problem is in the `catch(Exception $e)`, as it should be `catch(Throwable $e)`. The catch can be changed to fix the issue, but I don't like using `catch` unless it actually handles the error.

How about something like this?

```
    public function transactional(callable $callback)
    {
        $this->begin();

        $result = false;
        try {
            $result = $callback($this);
        } finally {
            if($result === false) {
                $this->rollback();
            } else {
                $this->commit();
            }
        }
        
        return $result;
    }
```

I think the above is easier to maintain and avoids the confusion between Exceptions and Throwables.

> note: catch(Exception $e) exists in a number of places in the code.

### PR DESCRIPTION
**Pull request by thinkingmedia**
Closes #8997 

A rollback will not be done in PHP7 if an Error type is thrown. Usage of a try/finally resolves this problem.

### CODE REVIEW

**Comment by thinkingmedia** [comment]
Travis produced a strange error for PHP 5.5, but everything else passed.

```
/home/travis/build.sh: line 45:  9503 Segmentation fault      (core dumped) vendor/bin/phpunit
```

Anyone know what this means?

**Comment by lorenzo** [comment]
It seems like this change hits a bug in php

**Comment by Magellol** [comment]
@thinkingmedia I've seen a couple of Seg faults when running tests on php 5.x in the past. I'm not really sure about its source but one thing for sure they fixed it in PHP7.

From what I've been searching in the past, it was something around memory loss.

**Comment by thinkingmedia** [comment]
@Magellol this might end up being to difficult to fix. I've found examples where people can cause it to happen with just 5 lines of code with PHP 5.2

But this is 5.5 so hopefully it's more stable than that.

I suspect that my change triggers a `rollback()` when PHP can't perform the operation and segs faults. Maybe I can add a check to see that the datasource is still connected or something like that.

**Comment by thinkingmedia** [comment]
I want to change this:

```
    public function disableConstraints(callable $callback)
    {
        $this->disableForeignKeys();

        try {
            $result = $callback($this);
        } catch (Exception $e) {
            $this->enableForeignKeys();
            throw $e;
        }

        $this->enableForeignKeys();
        return $result;
    }
```

to this:

```
    public function disableConstraints(callable $callback)
    {
        $this->disableForeignKeys();

        try {
            return $callback($this);
        } finally {
            $this->enableForeignKeys();
        }
    }
```

But I'll wait until I resolve the PHP 5.5 issues. So this is a **todo** reminder for myself.

File: tests/TestCase/ORM/TableTest.php [hunk 1]
```diff
@@ -2495,6 +2501,9 @@ public function testAtomicSaveRollback()
      */
     public function testAtomicSaveRollbackOnFailure()
     {
+// mathew
+$this->markTestIncomplete('Skipping');
```
Comment (thinkingmedia) [inline tests/TestCase/ORM/TableTest.php]: bite me!

**Comment by thinkingmedia** [comment]
**DO NOT MERGE THIS PR**
---

I'm going to mess around with things to see if I can narrow the scope on these travis issues. Just for the curiosity of wanting to know why this is failing.

**Comment by raul338** [comment]
You can enable travis to build on your github account, so when you finish you push it upstream and avoid set the warnings here :smile:

**Comment by thinkingmedia** [comment]
@raul338 that's a good idea, I'll close this PR and come back later if I get it fixed.

**Pull request closed without merging**